File: apps/web/src/lib/utils/order-code.ts
/**
 * Display code for orders.
 *
 *  - Production codes are generated server-side by the `reserve_order_display_code(date)`
 *    Postgres function and stored in `orders.display_code`. Format: `MK-NNN` (3-digit
 *    base), expanded automatically to `MK-NNNN`, `MK-NNNNN`, etc. when the per-day pool
 *    fills. Unique per KL calendar day (not globally).
 *
 *  - `getOrderDisplayCode(order)` returns the stored code when present and falls back to
 *    the legacy FNV-1a daily hash for any row that predates the migration.
 *
 *  - `generateOrderDisplayCode(id, date?)` is kept for the fallback only. New code should
 *    prefer `getOrderDisplayCode`.
 */

// FNV-1a 32-bit hash constants (legacy fallback)
const FNV_PRIME = 0x01000193
const FNV_OFFSET_BASIS = 0x811c9dc5

function fnv1a(input: string): number {
  let hash = FNV_OFFSET_BASIS
  for (let i = 0; i < input.length; i++) {
    hash ^= input.charCodeAt(i)
    hash = Math.imul(hash, FNV_PRIME)
  }
  return hash >>> 0
}

function getKualaLumpurDateString(date?: Date): string {
  const d = date || new Date()
  const formatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'Asia/Kuala_Lumpur',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  })
  return formatter.format(d)
}

export function generateOrderDisplayCode(orderId: string, date?: Date): string {
  const dateStr = getKualaLumpurDateString(date)
  const hash = fnv1a(orderId + dateStr)
  const num = hash % 1000
  return `MK-${num.toString().padStart(3, '0')}`
}

export interface OrderDisplayInput {
  id: string
  display_code?: string | null
}

export function getOrderDisplayCode(order: OrderDisplayInput, date?: Date): string {
  if (order.display_code) return order.display_code
  return generateOrderDisplayCode(order.id, date)
}

File: apps/merchant/lib/core/utils/order_code.dart
import '../../generated/tables/orders.dart';

/// Display code for orders.
///
///  - Production codes are stored in `orders.display_code` (written by the
///    Postgres trigger `orders_assign_display_code`). Format: `MK-NNN` base,
///    auto-expands to `MK-NNNN`, etc. when a day's pool fills.
///  - [getOrderDisplayCode] prefers the stored code and falls back to the
///    legacy FNV-1a daily hash for rows without one.
///  - [generateOrderDisplayCode] is the legacy fallback kept only so callers
///    that have just an id (e.g. before the order row loads) can render
///    something. Do not use for new callers that have the row.
String getOrderDisplayCode(OrdersRow order, [DateTime? date]) {
  final stored = order.getField<String>('display_code');
  if (stored != null && stored.isNotEmpty) return stored;
  return generateOrderDisplayCode(order.id, date);
}

/// Legacy deterministic daily display code. Kept only as a fallback for
/// pre-migration rows and for callers that have only an order id.
String generateOrderDisplayCode(String orderId, [DateTime? date]) {
  final now = date ?? DateTime.now();
  // Kuala Lumpur is UTC+8
  final klDate = now.toUtc().add(const Duration(hours: 8));
  final dateStr = '${klDate.year}-${_pad2(klDate.month)}-${_pad2(klDate.day)}';
  final hash = _fnv1a(orderId + dateStr);
  final num_ = hash % 1000;
  return 'MK-${num_.toString().padLeft(3, '0')}';
}

int _fnv1a(String input) {
  const int fnvPrime = 0x01000193;
  const int offsetBasis = 0x811c9dc5;
  int hash = offsetBasis;
  for (int i = 0; i < input.length; i++) {
    hash ^= input.codeUnitAt(i);
    hash = (hash * fnvPrime) & 0xFFFFFFFF;
  }
  return hash;
}

String _pad2(int value) => value.toString().padLeft(2, '0');
